src/Microsoft.OData.Core/ODataParameterWriterCore.cs

```diff
@@ -816,5 +879,135 @@ private void ValidateTransition(ParameterWriterState newState)
                     throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataParameterWriterCore_ValidateTransition_UnreachableCodePath));
             }
         }
+
+        /// <summary>
+        /// Asynchronously catch any exception thrown by the action passed in; in the exception case move the writer into
+        /// state ExceptionThrown and then rethrow the exception.
+        /// </summary>
+        /// <param name="action">The delegate to execute asynchronously.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        private async Task InterceptExceptionAsync(Func<Task> action)
+        {
+            try
+            {
+                await action().ConfigureAwait(false);
+            }
+            catch
+            {
+                this.EnterErrorScope();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously catch any exception thrown by the action passed in; in the exception case move the writer into
+        /// state ExceptionThrown and then rethrow the exception.
+        /// </summary>
+        /// <param name="func">The delegate to execute asynchronously.</param>
+        /// <returns>A task that represents the asynchronous operation. 
+        /// The value of the TResult parameter contains a T instance.</returns>
+        private async Task<T> InterceptExceptionAsync<T>(Func<Task<T>> func)
+        {
+            try
+            {
+                return await func().ConfigureAwait(false);

```

Comment:
>return [](start = 16, length = 6)

return is necessary here?

Reply:
@xuzhg Yes, because we use this overload in scenarios like below where we need the return value:
```csharp
            ODataWriter resourceWriter = await this.InterceptExceptionAsync(
                () => this.CreateResourceWriterImplementationAsync(parameterName, itemTypeReference)).ConfigureAwait(false);
```
It's for that same reason why it accepts a parameter of type `Func<Task<T>> func` - as opposed to `Func<Task>` for the overload